core/src/main/java/org/infinispan/xsite/statetransfer/XSiteStateProviderImpl.java

```diff
@@ -206,47 +206,42 @@ public void run() {
 
             CompletableFutures.await(stateTransferLock.topologyFuture(minTopologyId));
 
-            final List<XSiteState> chunk = new ArrayList<>(chunkSize <= 0 ? DEFAULT_CHUNK_SIZE : chunkSize);
+            final List<XSiteState> chunk = new ArrayList<>(chunkSize);
 
             if (debug) {
                log.debugf("[X-Site State Transfer - %s] start DataContainer iteration", xSiteBackup.getSiteName());
             }
 
-            for (InternalCacheEntry ice : dataContainer) {
-               if (canceled) {
-                  log.debugf("[X-Site State Transfer - %s] State transfer canceled!", xSiteBackup.getSiteName());
-                  return;
-               }
-               if (chunkSize > 0 && chunk.size() == chunkSize) {
-                  try {
-                     sendFromSharedBuffer(xSiteBackup, chunk, this);
-                  } catch (Throwable t) {
-                     error = true;
-                     log.unableToSendXSiteState(xSiteBackup.getSiteName(), t);
-                     return;
-                  }
-                  chunk.clear();
-               }
-               if (shouldSendKey(ice.getKey())) {
-                  if (trace) {
-                     log.tracef("Added key '%s' to current chunk", ice.getKey());
-                  }
-                  chunk.add(XSiteState.fromDataContainer(ice));
-               }
+            try {
+               Flowable.fromIterable(dataContainer)
+                       .flatMapMaybe(ice -> {
```

Comment:
I would prefer seeing `filter` and `map` here to avoid all the Maybe allocations.

Reply:
Done